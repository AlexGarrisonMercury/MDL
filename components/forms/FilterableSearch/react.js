import React, { PureComponent } from 'react'
import classNames from 'classnames/bind'
import { mChevronDown } from '@masonite/svg-icons'
import { React as Input } from 'components/forms/Input'
import { React as Checkbox } from 'components/forms/Checkbox'
import PropTypes from 'prop-types'
import style from './style.module.scss'

const cx = classNames.bind(style)

export default class FilterableSearch extends PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      showFilters: false,
    }

    this.dropdownMenuRef = React.createRef()
  }

  componentDidMount() {
    document.body.addEventListener('click', this.handleClick)
  }

  componentWillUnmount() {
    document.body.removeEventListener('click', this.handleClick)
  }

  handleClick = ({ target }) => {
    const { showFilters } = this.state
    const dropdownMenu = this.dropdownMenuRef.current

    if (showFilters && !dropdownMenu.contains(target)) {
      this.setState({ showFilters: false })
    }
  }

  render() {
    const {
      onFilterChange,
      onInputChange,
      onSearch,
      filterChoices = [],
      placeholder,
      value,
    } = this.props

    const { showFilters } = this.state

    const SearchButton = () => (
      <button
        type="button"
        onClick={() => onSearch(value)}
        className={cx(['FilterableSearch__btn'])}
      >
        Search
      </button>
    )

    const DropdownToggle = () => (
      <button
        type="button"
        className={cx(['FilterableSearch__dropdown-toggle'])}
        onClick={() => this.setState(prevState => ({
          showFilters: !prevState.showFilters,
        }))
        }
      >
        <div
          className={cx(['FilterableSearch__dropdown-toggle-icon'])}
          dangerouslySetInnerHTML={{ __html: mChevronDown }}
        />
      </button>
    )

    const DropdownMenu = ({ show }) => (
      <ul
        className={cx([
          'FilterableSearch__dropdown-menu',
          { 'FilterableSearch__dropdown-menu--open': show },
        ])}
      >
        {filterChoices.map(({ label, value, checked }, index) => (
          <li className={cx(['FilterableSearch__dropdown-menu-item'])} key={value}>
            {/* eslint-disable-next-line */}
            <label
              className={cx(['FilterableSearch__dropdown-menu-item-label'])}
              htmlFor={`${value}${index}`}
            >
              <Checkbox
                name={`${value}${index}`}
                checked={checked}
                onChange={() => onFilterChange({ label, value, checked })}
              />
              <span className={cx(['FilterableSearch__dropdown-menu-item-label-text'])}>
                {label}
              </span>
            </label>
          </li>
        ))}
      </ul>
    )

    return (
      <div className={cx(['FilterableSearch'])}>
        <Input
          placeholder={placeholder}
          onChange={value => onInputChange(value)}
          onKeyDown={ev => {
            if (ev.keyCode === 13) {
              onSearch(ev.target.value)
            }
          }}
          value={value}
        />
        <SearchButton />
        <div ref={this.dropdownMenuRef}>
          <DropdownToggle />
          <DropdownMenu show={showFilters} />
        </div>
      </div>
    )
  }
}

FilterableSearch.propTypes = {
  onFilterChange: PropTypes.func.isRequired,
  onInputChange: PropTypes.func,
  onSearch: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  filterChoices: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      value: PropTypes.string,
      checked: PropTypes.bool,
    }),
  ).isRequired,
  placeholder: PropTypes.string,
}

FilterableSearch.defaultProps = {
  placeholder: 'Search',
  onInputChange: () => {},
}
