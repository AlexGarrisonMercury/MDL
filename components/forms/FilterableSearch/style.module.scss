@import "styles/index";

.filterable-search {
  position: relative;
  font-family: $font-family;
  display: flex;
  align-items: top;
  width: 100%;

  $height: 36px;

  &__btn {
    display: inline-block;
    height: $height;
    border: 0;
    font-family: $font-family;
    font-size: 12px;
    color: $color-night;
    font-weight: 700;
    padding: $size-md;
    background: $color-limerick;
    cursor: pointer;
  }

  &__dropdown-toggle {
    @include square($height);
    display: inline-flex;
    border: 0;
    align-items: center;
    background: darken($color-limerick, 3%);
    color: $color-night;
    border-radius: 0 2px 2px 0;
    cursor: pointer;

    &--active {
      background: darken($color-limerick, 5%);
    }
  }

  &__dropdown-toggle-icon {
    @include square($size-lg);
    fill: $color-night;
    margin: auto;
    pointer-events: none;
    transform: translate3d(0, 0, 0);

    &--active {
      transform: translate3d(0, 1px, 0);
    }
  }

  &__dropdown-menu {
    position: absolute;
    right: 0;
    margin: 0;
    padding: $size-lg;
    background: $color-white;
    box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.15);
    list-style: none;
    pointer-events: none;
    opacity: 0;
    z-index: 10;

    &--open {
      opacity: 1;
      pointer-events: all;
    }
  }

  &__dropdown-menu-item {
    margin: 0 0 $size-sm;
    padding: 0;

    &:last-of-type {
      margin-bottom: 0;
    }
  }

  &__dropdown-menu-item-label {
    display: flex;
    align-items: center;
  }

  &__dropdown-menu-item-label-text {
    margin-left: $size-md;
  }
}
