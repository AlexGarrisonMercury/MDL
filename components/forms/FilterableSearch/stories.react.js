import React from 'react'
import { storiesOf } from '@storybook/react'
import { withKnobs } from '@storybook/addon-knobs'
import { withState } from '@dump247/storybook-state'
import FilterableSearch from './react'

storiesOf('Forms/FilterableSearch', module)
  .addDecorator(withKnobs)
  .add(
    'default',
    withState({
      filterChoices: [
        { label: 'Filter 1', value: 'FILTER1', checked: true },
        { label: 'Filter 2', value: 'FILTER2', checked: false },
      ],
    })(({ store }) => (
      <FilterableSearch
        handleFilterChange={checkedChoice => {
          store.set({
            ...store.state,
            filterChoices: store.state.filterChoices.map(choice => (choice.value === checkedChoice ? { ...choice, checked: !choice.checked } : choice)),
          })
        }}
        handleSearch={value => {
          console.log('handleSearch callback invoked, returns:', value)
        }}
        filterChoices={store.state.filterChoices}
      />
    )),
  )
