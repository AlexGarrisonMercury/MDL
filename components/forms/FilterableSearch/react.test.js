import React from 'react'
import { mount } from 'enzyme'
import { JSDOM } from 'jsdom'
import FilterableSearch from './react'

const jsdom = new JSDOM('<!doctype html><html><body><div id="app">Mount here</div></body></html>')
const { window } = jsdom

describe('FilterableSearch', () => {
  let props
  let wrapper

  beforeEach(() => {
    props = {
      onFilterChange: jest.fn(),
      onInputChange: jest.fn(),
      onSubmit: jest.fn(),
      value: '',
      filterChoices: [
        { label: 'Filter 1', value: 'FILTER1', checked: false },
        { label: 'Filter 2', value: 'FILTER2', checked: false },
      ],
    }

    wrapper = mount(<FilterableSearch {...props} />, {
      attachTo: window.document.querySelector('#app'),
    })
  })

  describe('with snapshots', () => {
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('with user actions', () => {
    it('should display the filter options when clicking the dropdown button', () => {
      expect(wrapper.find('DropdownMenu').prop('show')).toEqual(false)
      wrapper.find('DropdownToggle').simulate('click')
      expect(wrapper.find('DropdownMenu').prop('show')).toEqual(true)
      wrapper.find('DropdownToggle').simulate('click')
      expect(wrapper.find('DropdownMenu').prop('show')).toEqual(false)
    })
    it('should trigger search action when the user hits enter in input or clicks SearchButton', () => {
      expect(props.onSubmit).toBeCalledTimes(0)
      wrapper.find('SearchButton').simulate('click')
      expect(props.onSubmit).toBeCalledTimes(1)
      wrapper.find('.text-field__input').simulate('keyDown', { keyCode: 1 }) // not Enter key
      wrapper.find('.text-field__input').simulate('keyDown', { keyCode: 13 }) // Enter key
      expect(props.onSubmit).toBeCalledTimes(2)
    })
  })
})
