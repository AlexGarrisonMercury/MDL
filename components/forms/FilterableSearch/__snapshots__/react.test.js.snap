// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterableSearch with snapshots should render correctly 1`] = `
<FilterableSearch
  filterChoices={
    Array [
      Object {
        "checked": false,
        "label": "Filter 1",
        "value": "FILTER1",
      },
      Object {
        "checked": false,
        "label": "Filter 2",
        "value": "FILTER2",
      },
    ]
  }
  handleFilterChange={[MockFunction]}
  handleSearch={[MockFunction]}
  placeholder="Search"
>
  <div
    className="FilterableSearch"
  >
    <Input
      error={false}
      onKeyDown={[Function]}
      placeholder="Search"
    >
      <div
        className="input-wrapper"
      >
        <input
          className="input"
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Search"
          type="text"
          value=""
        />
        <div
          className="input-icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\"><title>x</title><path d=\\"M13.4 12l5.3-5.3c0.4-0.4 0.4-1 0-1.4s-1-0.4-1.4 0l-5.3 5.3-5.3-5.3c-0.4-0.4-1-0.4-1.4 0s-0.4 1 0 1.4l5.3 5.3-5.3 5.3c-0.4 0.4-0.4 1 0 1.4 0.2 0.2 0.4 0.3 0.7 0.3s0.5-0.1 0.7-0.3l5.3-5.3 5.3 5.3c0.2 0.2 0.5 0.3 0.7 0.3s0.5-0.1 0.7-0.3c0.4-0.4 0.4-1 0-1.4l-5.3-5.3z\\"></path></svg>",
            }
          }
        />
      </div>
    </Input>
    <SearchButton>
      <button
        className="FilterableSearch__btn"
        onClick={[MockFunction]}
        type="button"
      >
        Search
      </button>
    </SearchButton>
    <div>
      <DropdownToggle>
        <button
          className="FilterableSearch__dropdown-toggle"
          onClick={[Function]}
          type="button"
        >
          <div
            className="FilterableSearch__dropdown-toggle-icon"
            dangerouslySetInnerHTML={
              Object {
                "__html": "<svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\"><title>chevron-down</title><path d=\\"M18.7 8.3c-0.4-0.4-1-0.4-1.4 0l-5.3 5.3-5.3-5.3c-0.4-0.4-1-0.4-1.4 0s-0.4 1 0 1.4l6 6c0.2 0.2 0.5 0.3 0.7 0.3s0.5-0.1 0.7-0.3l6-6c0.4-0.4 0.4-1 0-1.4z\\"></path></svg>",
              }
            }
          />
        </button>
      </DropdownToggle>
      <DropdownMenu
        show={false}
      >
        <ul
          className="FilterableSearch__dropdown-menu"
        >
          <li
            className="FilterableSearch__dropdown-menu-item"
            key="FILTER1"
          >
            <label
              className="FilterableSearch__dropdown-menu-item-label"
              htmlFor="FILTER10"
            >
              <Checkbox
                checked={false}
                name="FILTER10"
                onChange={[Function]}
              >
                <div
                  className="Checkbox"
                  onClick={[Function]}
                >
                  <input
                    checked={false}
                    className="Checkbox__checkbox"
                    readOnly={true}
                    type="checkbox"
                  />
                  <div
                    className="Checkbox__svg-wrapper"
                  >
                    <svg
                      className="Checkbox__svg-icon Checkbox__svg-icon--checked"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "<svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\"><title>check</title><path d=\\"M20.7 5.3c-0.4-0.4-1-0.4-1.4 0l-10.3 10.3-4.3-4.3c-0.4-0.4-1-0.4-1.4 0s-0.4 1 0 1.4l5 5c0.2 0.2 0.4 0.3 0.7 0.3s0.5-0.1 0.7-0.3l11-11c0.4-0.4 0.4-1 0-1.4z\\"></path></svg>",
                        }
                      }
                    />
                  </div>
                </div>
              </Checkbox>
              <span
                className="FilterableSearch__dropdown-menu-item-label-text"
              >
                Filter 1
              </span>
            </label>
          </li>
          <li
            className="FilterableSearch__dropdown-menu-item"
            key="FILTER2"
          >
            <label
              className="FilterableSearch__dropdown-menu-item-label"
              htmlFor="FILTER21"
            >
              <Checkbox
                checked={false}
                name="FILTER21"
                onChange={[Function]}
              >
                <div
                  className="Checkbox"
                  onClick={[Function]}
                >
                  <input
                    checked={false}
                    className="Checkbox__checkbox"
                    readOnly={true}
                    type="checkbox"
                  />
                  <div
                    className="Checkbox__svg-wrapper"
                  >
                    <svg
                      className="Checkbox__svg-icon Checkbox__svg-icon--checked"
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "<svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\"><title>check</title><path d=\\"M20.7 5.3c-0.4-0.4-1-0.4-1.4 0l-10.3 10.3-4.3-4.3c-0.4-0.4-1-0.4-1.4 0s-0.4 1 0 1.4l5 5c0.2 0.2 0.4 0.3 0.7 0.3s0.5-0.1 0.7-0.3l11-11c0.4-0.4 0.4-1 0-1.4z\\"></path></svg>",
                        }
                      }
                    />
                  </div>
                </div>
              </Checkbox>
              <span
                className="FilterableSearch__dropdown-menu-item-label-text"
              >
                Filter 2
              </span>
            </label>
          </li>
        </ul>
      </DropdownMenu>
    </div>
  </div>
</FilterableSearch>
`;
